src/wrappers/themis/Obj-C/objcthemis/smessage.m

```diff
@@ -138,12 +142,35 @@ - (nullable NSData *)wrapData:(nullable NSData *)message error:(NSError * __auto
 
 - (nullable NSData *)unwrapData:(nullable NSData *)message error:(NSError * __autoreleasing *)error {
     size_t unwrappedMessageLength = 0;
-
-    TSErrorType result = (TSErrorType) themis_secure_message_unwrap([self.privateKey bytes], [self.privateKey length],
-        [self.publicKey bytes], [self.publicKey length], [message bytes], [message length],
-        NULL, &unwrappedMessageLength);
+    TSErrorType result = TSErrorTypeFail;
+  
+    switch (self.mode) {
+      case TSMessageModeEncryptDecrypt:
+        result = (TSErrorType) themis_secure_message_decrypt([self.privateKey bytes], [self.privateKey length],
+                                                             [self.publicKey bytes], [self.publicKey length], [message bytes], [message length],
+                                                             NULL, &unwrappedMessageLength);
+        break;
+        
+      case TSMessageModeSignVerify:
+        if (!(self.publicKey) || [self.publicKey length] == 0) {
+          NSLog(@"Error during verifying: public key is missing");
+          if (error) {
+            *error = SCERROR(TSErrorTypeFail, @"Secure Message failed wraping");
```

Comment:
I guess we can update NSError descriptions as well (here an in other places), if we add logs and perform other refactoring:

```suggestion
            *error = SCERROR(TSErrorTypeFail, @"Secure Message failed to verify message");
```

Reply:
agreed, updated